### PR TITLE
ISO5457 worksheets for KiCad

### PR DESCRIPTION
**Pull request by erasmo1982**
Re-push of https://github.com/KiCad/kicad-templates/pull/13

### CODE REVIEW

**Comment by antoniovazquezblanco** [comment]
Assigned to @evanshultz because the original PR was also.

**Comment by poeschlr** [comment]
Is it on purpose that some of the frame separators interfere with information fields? Seems to affect all A4 layouts.

Example:
Screenshot from 2019-09-04 18-49-34

**Comment by erasmo1982** [comment]
These elements are the centring marks. There should be 4 centring marks. It is recommended to represent them as continuous lines, with length 10 mm and width 0.7 mm, however their form may be chosen freely. As far as I know there is not any exception for A4 paper size.

**Comment by poeschlr** [comment]
I fear these standards are mainly made for larger paper sizes (This is why they take up this much room of an A4 page).

So it might be a good idea to use some common sense and at least remove the bottom one. (you could replace it by one that has double the thickness for example)

**Comment by erasmo1982** [comment]
It is not a bad idea. I will update the A4 worksheets and push the changes in the coming days.

**Comment by erasmo1982** [comment]
Centring marks in A4 worksheets have been shortened. Worksheets for other paper sizes have not been modified.

This is the result:
image

**Comment by poeschlr** [comment]
@evanshultz this would be ok for me but i have not looked into the standard. Could you take a quick look?
I hope my decision to not follow the centering mark standard for A4 is ok for you.

**Comment by evanshultz** [comment]
The above looks good to me. I have the standard and really wanted to review these and get them merged but I failed earlier. Please give me a couple days (DirectFET footprints and KLC website updates first) to get around to this.

@erasmo1982 
Thanks for your diligence and fast responses!

**Comment by chschlue** [comment]
Looks good.

Could you maybe change rev to ASMEY1435M or something like that? That would make it easier to add other variants in the future.

**Comment by erasmo1982** [comment]
Sure. I will update the revision block according to ASME Y14.35M. I plan to commit the changes in a few days.

**Comment by chschlue** [comment]
Oh, that's different from my intention but also more than I asked for :)
I was actually just thinking about file names so possible future variants with revision blocks following other standards wouldn't have to be named _rev2 and so on.

**Comment by erasmo1982** [comment]
Revision block has been updated according to ASME Y14.35M. In addition, files have been renamed according to the corresponding standards.

**Comment by chschlue** [comment]
Title block conforms to ISO 7200:2004

**Comment by chschlue** [comment]
Bad news on ISO 5457:1999 + A1:2010:
- [ ] A4 grid references _must_ be on top and right edges (4.4)
- [ ] Grid _must_ be centered on the sheet, not on the drawing area (4.4)
- [ ] Same for center marks (4.3)
- [ ] Center marks _should_ extend 10mm into the drawing area (for A4, this has already been discussed)

**Comment by erasmo1982** [comment]
Thanks for the observation! I will fix this.

**Comment by chschlue** [comment]
@evanshultz 
The German ISO 5457 seems to say centring marks should extend 10mm into the drawing area, making them 15mm long while fig. 5 suggests they extend 5mm, 10mm total.
Can you check the English version?

**Comment by cpresser** [comment]
For reference:
image

**Comment by chschlue** [comment]
Yes, I read that. The Austrian and German translations seem to match :)

Stupid me. Centring marks in figure 5 are dimensioned, so my interpretation of section 4.3 was right.

**Comment by chschlue** [comment]
That this is too much for A4 seems to be consensus.
How about 5mm for A3 and the recommended 10mm for A2 (and larger, later maybe)?

**Comment by chschlue** [comment]
kmmm-004

**Comment by erasmo1982** [comment]
Examples I have found suggest that centring lines are 10 mm long.

**Comment by chschlue** [comment]
That's not wrong, the standard just recommends 15mm after all.
I just think following the recommendation won't hurt on larger sheets.

**Comment by evanshultz** [comment]
@poeschlr 
Sure. I will be able to do it on Monday.

**Comment by chschlue** [comment]
One more thing I noticed:
%S/%N text is too wide for double digit sheet numbers.

**Comment by erasmo1982** [comment]
It is true. I will fix it.

**Comment by evanshultz** [comment]
1. With my cursor at the bottom right corner, and with the top left corner at 20x10mm, I find the A4 drawing space area to be 267x189mm. Should be 277x180mm. Is this my problem? (The portrait A4 sheets look OK to me.)
image

2. The largest block of text, showing 'Legal owner' in ISO7200, seems to have a much bigger text size than that shown in the standard. Should this text be shrunken a bit? Relatedly, if I enter 'Sub-assembly drawing' into Comment2 that text just fits. It's 20 characters. ISO7200 says that field is recommended to support 30 characters. Because the font in KiCad is not fixed width this is a bit hard to pin down. How do we want to handle this? Just use English, A though whatever character we end up with, and adjust the text size to fit the ISO7200 recommended character length?

Too bad we don't get justification and text size support for variables in the title block. Or custom variable extraction from the title block. That would be a feature request for the devs.

**Comment by evanshultz** [comment]
And some things I wrote at https://github.com/KiCad/kicad-templates/pull/13#issuecomment-398188629 that seem to still be relevant:
1. Should we include the year of release (version) of the standards like ISO7200:2004 in the worksheet file names? I think yes. What about when a future ISO standard is released and we want to differentiate the new one from this one?
1. Can you fully write out `compact` in the file names for clarity? Not that it's impossible to understand as it is now but I don't see a drawback and it may help.
1. http://tolerancing.net/engineering-drawing/revisions-block.html does not show the ZONE column in the rev history. I don't have ASME Y14.35M. Is that column in the real standard? I see it in 14.100 but it's always blank and I didn't see a justification for having it.

**Comment by chschlue** [comment]
> 1. With my cursor at the bottom right corner, and with the top left corner at 20x10mm, I find the A4 drawing space area to be 267x189mm. Should be 277x180mm. Is this my problem? (The portrait A4 sheets look OK to me.)

This must be an oversight in the standard when Amd.1:2010 introduced A4 landscape.
Table 1 still only gives the drawing area dimensions for A4 portrait.

A4 portrait drawing area is 20mm from one _long_ edge, 10mm from the others -> 180x277mm
A4 landscape is 20mm from one _short_ edge, 10mm from the others -> 267x190mm

**Comment by chschlue** [comment]
> 2. The largest block of text, showing 'Legal owner' in ISO7200, seems to have a much bigger text size than that shown in the standard. Should this text be shrunken a bit? Relatedly, if I enter 'Sub-assembly drawing' into Comment2 that text just fits. It's 20 characters. ISO7200 says that field is recommended to support 30 characters. Because the font in KiCad is not fixed width this is a bit hard to pin down. How do we want to handle this? Just use English, A though whatever character we end up with, and adjust the text size to fit the ISO7200 recommended character length?

Not sure what ISO7200's authors were actually up to.
Table 2 footnote a suggests they are actually talking about `bytes` or `octets`, not `glyphs` or `letters`.

**Comment by chschlue** [comment]
> 1. Should we include the year of release (version) of the standards like ISO7200:2004 in the worksheet file names? I think yes. What about when a future ISO standard is released and we want to differentiate the new one from this one?

Any formatting suggestions? Colons are out of the question and something like A2_ISO7200_2004_ISO5457_1999_Amd1_2010_ASMEY1435M_2015.kicad_wks is not very palatable.

**Comment by evanshultz** [comment]
> > 1. The largest block of text, showing 'Legal owner' in ISO7200, seems to have a much bigger text size than that shown in the standard. Should this text be shrunken a bit? Relatedly, if I enter 'Sub-assembly drawing' into Comment2 that text just fits. It's 20 characters. ISO7200 says that field is recommended to support 30 characters. Because the font in KiCad is not fixed width this is a bit hard to pin down. How do we want to handle this? Just use English, A though whatever character we end up with, and adjust the text size to fit the ISO7200 recommended character length?
> 
> Not sure what ISO7200's authors were actually up to.
> Table 2 footnote a suggests they are actually talking about `bytes` or `octets`, not `glyphs` or `letters`.

Ahh. I see. They do use the word 'characters', and then show an extended length for CJK. So how do we interpret this? I assume we should support all worldwide languages, but if the support is based on the length of the bytes instead of character I'm lost. If this is about rendered characters, it also doesn't mention kerning which would have a notable effect over 25-30 characters.

Are you suggesting we leave it as is? If the 'Legal owner' text fills up the block and we can't use line breaks it seems like too big of a font size...

**Comment by evanshultz** [comment]
> Any formatting suggestions? Colons are out of the question and something like A2_ISO7200_2004_ISO5457_1999_Amd1_2010_ASMEY1435M_2015.kicad_wks is not very palatable.

This can get ugly. What about hyphens to replace colons and underscores to separate the section. So `<page size>_<drawing sheet standard>-<drawing sheet standard year>_<title block standard>-<title block standard year>`?

I didn't include ASME Y14.35M here because, well, it made me wonder if there an ISO standard for rev history? Are we mixing things together and ASME is its own worksheet?

**Comment by chschlue** [comment]
All the quotes make this conversation hard to follow, I'll try without them.

Re A4 landscape drawing area size: does my comment make sense?

Re field size: I have the German version, so I'm not sure what simply got lost in translation. My interpretation is in fact bytes, making that part irrelevant here. If they actually mean rendered characters, I'm pretty sure they would have been anal enough to reference the ISO 3098 font and specify character width, height, line thickness, everything.

Re file name: Still ugly, but I don't have any better idea. OTOH, what's the actual benefit of including the year? When the standard is updated, the template should be updated as well.
Standards bodies tend to be light on breaking changes or assign a new document number in case it's a complete complete overhaul.

**Comment by chschlue** [comment]
Re ISO rev history: Couldn't find anything. Can't believe it.

I'm not familiar with ASME standards but I don't think they'd only standardize a revision block and nothing else.

**Comment by evanshultz** [comment]
Maybe let's try enumeration. I redacted your last comment.

1. Yes. Is that in the standard anywhere? I don't see it in the 1999 version.

2. Below is the 2004 English version. The word 'characters' doesn't make me think of anything but rendered glyphs, but I would guess you are correct. And I'm really not sure how the number of bytes a character consumes is relevant for a title block; isn't only the rendered version of the language and shown on screen or when printed important here? I reckon as long as we can fit a reasonable number of characters in the field we're OK, and overall it generally feels a bit tight to me.
image

3. My consideration is being able to support all various worksheet versions in KiCad at once for companies or people that may not use the latest version for whatever reason. I agree there probably isn't going to be a big change across versions but you never know. And it seems like the cautious thing to prepare for the future in whatever sense that occurs.

4. Crazy, isn't it? No rev history document? ASME appears to have everything but I'm not too familiar either. Check out Y14.100. At least a title block is shown there (but not at the level of detail of ISO7200).

**Comment by chschlue** [comment]
1. ISO 5457:1999 only specifies A[0-3] landscape and A4 portrait. The 2010 amendment adds A4 landscape. This doesn't matter for figures 1 and 2, but table 2 would have had to be updated and was apparently forgotten. A4 a1 and b1 are 210mm and 297mm, respectively, but a2 and b2 change with orientation.
iso5457-000
iso5457-001

2. Adding five bytes to support two-byte encodings doesn't make much sense to me but adding five glyphs to support Japanese and Chinese makes no sense whatsoever. I'm lost.
It's a recommendation after all, so let's just agree on shrinking the text a little.

3. I think we can agree on keeping them somewhere (other than a git commit that is), but how about an `obsolete` subdirectory? We have obsolete symbols although there are still people using Nixie tubes...

4. The ISO-ASME-Frankensheets are not ideal, but they don't block full-ISO or full-ASME additions, so why not?

**Comment by evanshultz** [comment]
1. Got it.
1. Sounds good to me.
1. Again playing Devil's advocate... what about when we have two old versions? Are there multiple obsolete directories, or multiple sub-directories under obsolete? Or we can merge like this and then deal with it later if and when that happens.
1. I'm fine with it. Just wanted to check if ISO had their own rev history doc. That was answered.

**Comment by chschlue** [comment]
3. If the path changes to obsolete/..., users will have to update their projects manually anyway, so the year could be added to the filename then, too.
I really don't like the idea of just piling up obsolete files in Worksheets/
Hmm... how about adding the year now but still moving them when a new version is available? Might at least be a halfway clean solution.
4. -
5. Any thoughts about https://github.com/KiCad/kicad-templates/pull/25#issuecomment-531485658
6. With already ugly filenames, how about appending `_<ISO 639-1:2002 language code>`?

**Comment by erasmo1982** [comment]
I have decreased the text size in title boxes. This feature improves readability and increases the allowed number of characters.

**Comment by evanshultz** [comment]
@erasmo1982 Thanks for your continued work and patience.
Would you be able to look at items 2 and 3 I left at https://github.com/KiCad/kicad-templates/pull/25#issuecomment-536096220?

@chschlue 
3 Yes, that is what I was thinking. @erasmo1982 would you mind making a change like discussed above?
5 I guess you're talking about the centering marks? I only have the 1999 version and it is also confusing. It says they should extend 10mm beyond the drawing frame. But in fig 5, they are clearly only 10mm long (double the grid reference border width). Going by the assumption that the human eye is easier to understand images than text, I guess fig 5 is correct and 10mm long lines 5mm into the drawings area is the intention. Cutting back the A4 lines works for me as Rene suggested.
6 What did you have in mind for this to indicate? The language of the text in the title block, right? If so, that's fine with me. Users could then download this and customize the characters while keeping this standardized format and title block. (Interestingly, I can type Chinese characters into the Page Settings dialog but they render as `?` in the title block.)

**Comment by chschlue** [comment]
3. ✓
4. --
5. ✓
6. Exactly.
Chinese gives me question marks as well but Cyrillic works for me. I even tried runes, which give me boxes, indicating they're missing from the font but would otherwise work.

**Comment by erasmo1982** [comment]
Hello. I think it is not necessary to have an obsolete directory for templates. Older ones may be found in the repository. Respect to the centring marks, I prefer to leave them as they are, as longer centring marks are more intrusive.

**Comment by evanshultz** [comment]
@erasmo1982 
Yes, we all agree on centering marks.

Can you please add the revision year for the standards and also the language code to the file names as mentioned above? I believe that is the only thing to do.

**Comment by erasmo1982** [comment]
Sure! I plan to do it in the next few days.

**Comment by evanshultz** [comment]
@erasmo1982 
Thanks!

Can anyone verify the `ASMEY1435M-2015` part of the files? Does the rev history block conform to that doc from that year?

**Comment by chschlue** [comment]
I don't have access to the full text but it looks like the current revision is ASME Y14.35 2014 (no M and different year).

**Comment by erasmo1982** [comment]
@evanshultz @chschlue do you consider everything is ok in this branch to be merged?

**Comment by chschlue** [comment]
Like I said, I can't verify the revision block (but it appears to be correct).

IMO, this can be merged.
Waiting for a mechanical engineer with ASME standards on hand to pass by an ECAD repo could take forever, after all :)

**Comment by evanshultz** [comment]
OK. Let's do it. Any final objections? If this isn't merged within 24 hours I'll merge it.

Huge thanks to @erasmo1982 for taking the initiative on this and for all the work over such a long period of time to get this done!

**Comment by chschlue** [comment]
I think everyone is just waiting for you to merge, you're assigned to it after all :)

**Comment by chschlue** [comment]
🥳
Thanks, @erasmo1982

**Comment by erasmo1982** [comment]
Good job guys! :)
